File: files/code/algorithms/java/SDRW/src/edu/uts/aai/utils/DSVL.java
/*
* To change this license header, choose License Headers in Project Properties.
* To change this template file, choose Tools | Templates
* and open the template in the editor.
*/
package edu.uts.aai.utils;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Hashtable;
import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;
import weka.core.Attribute;
import weka.core.FastVector;
import weka.core.Instance;
import weka.core.Instances;
import weka.filters.Filter;
import weka.filters.unsupervised.attribute.Remove;

/**
 * This class is the main class of the DSFS algorithm, which consists of three components,
 * including intra-feature value outlierness computing (i.e., the delta function in our ICDM2016 paper),
 * the adjacent matrix of the feature graph, and dense subgraph discovery of the feature graph.
 * Note that we skip the value graph construction and go directly to construct the feature graph,
 * which can speed up the algorithm a bit.
 * @author Guansong Pang
 */
public class DSVL {
    
    private ArrayList<ValueCentroid> cpList;
    
    private ArrayList<ArrayList<Double>> vMatrix = new ArrayList<ArrayList<Double>>();
    
    private ArrayList<Integer> valFeatIDs = new ArrayList<Integer>();  //to store feature id to each feature value
    private ArrayList<String> valIdentifier = new ArrayList<String>();
    private ArrayList<Double> valIndWgts = new ArrayList<Double>();   //to store total weights for each feature value
    
    
    private int dim;
    private double[] attrWeight; //to store the weights of all the features
    /**
     *
     * @param cpList the list of coupled centroids: each centroid contains the co-occurrence frequency of each value with other values
     */
    public DSVL(ArrayList<ValueCentroid> cpList) {
        this.cpList = cpList;
        this.dim = cpList.size();
    }
    
    /**
     * the main method for calling Charikar greedy, sequential backward and Las Vegas based
     * dense subgraph discovery
     * @return the IDs of features to be removed
     */
    public String denseSubgraphDiscovery(int dataSize) {
        calcIntraFeatureValueOutlierness(dataSize);
//        adjacentMatrix(cpList);
        valueAdjacencyMatrix();
        ArrayList<String> discardVals = new ArrayList<String>();
        double[] den = charikarGreedySearchforValueGraph(discardVals);
        double max = Double.MIN_VALUE;
        int maxID = -1;
        for(int i = 0; i < discardVals.size(); i++) {
            if(den[i] > max) {
                max = den[i];
                maxID = i;
            }
        }
        System.out.println("MAX:"+(new DecimalFormat("#0.000000")).format(max)+" ");
        Plot.plotYPoints(den, 3, DSVL4ODUtils.dataSetName, DSVL4ODUtils.dataSetName, "Iteration", "Avg. Incoming Edge Weight");
        return discardVals.get(maxID);
    }
    
    public void valueOutliernessLearning(int dataSize) {
        calcIntraFeatureValueOutlierness(dataSize);
        valueAdjacencyMatrix();
        ArrayList<String> discardVals = new ArrayList<String>();
        charikarGreedySearchforValueGraph(discardVals);
        weightedDegree2Score();
    }
    
    /**
     * to calculate the outlierness of each feature value based on the extent the value frequent deviating from the mode frequency
     * @param dataSize the number of instances in the data set
     */
    public void calcIntraFeatureValueOutlierness(int dataSize) {
        double [] mFreq = new double[dim];
        for(int i = 0; i < cpList.size(); i++) {
            ValueCentroid cp = cpList.get(i);
            int len = cp.getCenList().size();
            double maxFreq = 0;
            for(int j = 0; j < len; j++) {
                ValueNode cen = cp.getCenList().get(j);
                double globalFreq = cen.globalFreq(i, j);
                if(globalFreq > maxFreq)
                    maxFreq = globalFreq;
            }
            mFreq[i] = maxFreq;
        }
        
        double max = Double.MIN_VALUE;
        double min = Double.MAX_VALUE;
        for(int i = 0; i < cpList.size(); i++) {
            ValueCentroid cp = cpList.get(i);
            int len = cp.getCenList().size();
            int count = 0;
            for(int j = 0; j < len; j++) {
                ValueNode cen = cp.getCenList().get(j);
                double globalFreq = cen.globalFreq(i, j);
                if(globalFreq == 0) {
                    continue;
                }
                double intra;
                intra = (Math.abs(globalFreq-mFreq[i])/mFreq[i] + (1-mFreq[i]/dataSize))/2.0; //mode absolute difference based. '1.0/dataSize' is used to avoid zero outlierness
                cen.setIntraOD(intra);
                count++;
            }
        }
    }
    /**
     * to generate the adjacent matrix for the value-value graph
     */
    public void valueAdjacencyMatrix() {
//        int nonzeroVal = 0;
        System.out.println();
        for(int i = 0; i < dim; i++) {
            ValueCentroid cp = cpList.get(i);
            int count = 0;
            for(int j = 0; j < cp.getCenList().size(); j++) {
                ValueNode cen = cp.getCenList().get(j);
                if(cen.globalFreq(i, j) == 0) { //skip zero-apperance values
                    
                    System.out.print("0.0"+" ");
                    continue;
                }
                ArrayList<Double> col = new ArrayList<Double>();
                double tmp=0;
                for(int k = 0; k < dim; k++) {
                    FeatureInfo ai = cen.getAttrList().get(k); //contain local co-occurrence informaiton, i.e., co-occurrnce frequency of the i-th and k-th feature values
                    FeatureInfo gai = cp.getGlobalCentroid().getAttrList().get(k); //contain global co-occurrencd information
                    int len = ai.NumofValue();
                    for(int l = 0; l < len; l++) {
                        if (k == cen.getOrgFeat() && gai.value(l) != 0) { //skip zero-appearance values
//                            if(j == l) {
//                                col.add(cen.getIntraOD());  // self-loop value graph
//                                tmp += cen.getIntraOD();
//                            }
//                            else
                            col.add(0.0);
                            System.out.print("0.0"+" ");
                            continue;
                        }/**/
                        double freq = ai.value(l);
                        double gFreq = gai.value(l);
                        double cenFreq = cen.globalFreq(i, j);
                        if(cenFreq != 0 && gFreq != 0) { //skip zero-appearance values
//                            double w = cen.getIntraOD() * cpList.get(k).getCenList().get(l).getIntraOD() * Math.log10(freq*1.0/(cenFreq * gFreq))/Math.log10(2.0);
                            double w = cen.getIntraOD() * cpList.get(k).getCenList().get(l).getIntraOD() * (freq*1.0/(cenFreq * gFreq));
//                            double w = cen.getIntraOD() * cpList.get(k).getCenList().get(l).getIntraOD(); //DSVLia
//                            double w =  (freq*1.0/(cenFreq * gFreq));  //DSVLie
                            
                            col.add(w);
                            tmp += w;
                            System.out.print(w+" ");
                        } else {
                            System.out.print("0.0"+" ");
                        }
                    }
                }
                vMatrix.add(col);
                //  valIndWgts.add(tmp);
                valFeatIDs.add(i);
                valIndWgts.add(tmp);
                count++;
                valIdentifier.add(i+"_"+j);
                cen.setWeightedDegree(tmp);
                System.out.println();
            }
        }
    }
    
    public void updateWeightedDegree() {
        for(int i = 0; i < dim; i++) {
            ValueCentroid cp = cpList.get(i);
            for(int j = 0; j < cp.getCenList().size(); j++) {
                ValueNode cen = cp.getCenList().get(j);
                double tmp = 0;
                if(cen.globalFreq(i, j) == 0) //skip zero-apperance values
                    continue;
                for(int k = 0; k < dim; k++) {
                    FeatureInfo ai = cen.getAttrList().get(k); //contain local co-occurrence informaiton, i.e., co-occurrnce frequency of the i-th and k-th feature values
                    FeatureInfo gai = cp.getGlobalCentroid().getAttrList().get(k); //contain global co-occurrencd information
                    int len = ai.NumofValue();
                    for(int l = 0; l < len; l++) {
                        if (k == cen.getOrgFeat() && gai.value(l) != 0) { //skip zero-appearance values
                            continue;
                        }
                        double freq = ai.value(l);
                        double gFreq = gai.value(l);
                        double cenFreq = cen.globalFreq(i, j);
                        if(cenFreq != 0 && gFreq != 0) { //skip zero-appearance values
//                            double w = cen.getIntraOD() * cpList.get(k).getCenList().get(l).getIntraOD() * Math.log10(freq*1.0/(cenFreq * gFreq))/Math.log10(2.0);
//                            double w = cen.getWeight() *  cpList.get(k).getCenList().get(l).getWeight(); // DSVLia
                            double w = cen.getWeight() *  cpList.get(k).getCenList().get(l).getWeight() //DSVLie and DSVL
                                    * (freq*1.0/(cenFreq * gFreq));
                            tmp += w;
                        }
                    }
                }
                cen.setWeightedDegree(tmp);
            }
        }
    }
    
    public void weightedDegree2Score() {
        double vol = 0;
        weightNormalization();
        updateWeightedDegree();
        for(int i = 0; i < dim; i++) {
            ValueCentroid cp = cpList.get(i);
            for(int j = 0; j < cp.getCenList().size(); j++) {
                ValueNode cen = cp.getCenList().get(j);
//                vol += cen.getWeightedDegree() * cen.getWeight();
                vol += cen.getWeightedDegree() ;
            }
        }
        
        attrWeight = new double[dim];
        for(int m=0; m < dim; m++) {
            attrWeight[m] = 1;
        }
        
        for(int i = 0; i < dim; i++) {
            ValueCentroid cp = cpList.get(i);
            for(int j = 0; j < cp.getCenList().size(); j++) {
                ValueNode cen = cp.getCenList().get(j);
//                double prob = cen.getWeightedDegree() * cen.getWeight() / vol;
//                double prob = cen.getWeightedDegree() / vol;
                double prob = cen.getWeight();
                cen.setOutlierScore(prob);
                System.out.print(prob+",");
                attrWeight[i] *= (1-prob); //compute attribute weight
            }
        }
        System.out.println();
        for(int m=0; m < dim; m++) {
            attrWeight[m] = 1-attrWeight[m];
        }
    }
    
    public void weightNormalization() {
        double totalWgt = 0;
        
        for(int i = 0; i < dim; i++) {
            ValueCentroid cp = cpList.get(i);
            for(int j = 0; j < cp.getCenList().size(); j++) {
                ValueNode cen = cp.getCenList().get(j);
                totalWgt += cen.getWeight();
            }
        }
        
        for(int i = 0; i < dim; i++) {
            ValueCentroid cp = cpList.get(i);
            for(int j = 0; j < cp.getCenList().size(); j++) {
                ValueNode cen = cp.getCenList().get(j);
                cen.setWeight(cen.getWeight() / totalWgt);
            }
        }
    }
    
    public Hashtable<Integer,Double> scoringTestInstances(Instances data) {        
        Hashtable<Integer,Double> scores = new Hashtable<Integer,Double>();        
        for(int j = 0; j < data.numInstances(); j++) {
            Instance inst = data.instance(j);
            double score=1;
            for(int i = 0; i <dim; i++) {
                ValueCentroid cp = cpList.get(i);
                int index = ((Double)inst.value(i)).intValue();
                ValueNode cen = cp.getCenList().get(index);
                double s = cen.getOutlierScore();
//                double s = cen.getOutlierDegree()* attrWeight[i]; // weighted sum
                //if(score < s)
                //   score = s;
//                score *= Math.pow(1-s,1);
                score *= Math.pow(1-s,attrWeight[i]);
//                score += s;
                
            }            
            score = 1-score;
            System.out.println(score);
            scores.put(j,score);
        }
        
        return scores;
    }
    /**
     * to search for the densest subgraph in indirected graphs by using Charikar's greedy method presented in the paper below
     * @incollection{charikar2000greedy,
     * title={Greedy approximation algorithms for finding dense components in a graph},
     *   author={Charikar, Moses},
     *   booktitle={Approximation Algorithms for Combinatorial Optimization},
     *   pages={84--95},
     *   year={2000},
     *   publisher={Springer}
     * }
     * @param discardFeats the list to store non-relevant feature ids
     * @return the density array that records all densities of all the subgraphs
     */
    public double[] charikarGreedySearchforValueGraph(ArrayList<String> discardVals) {
        int len = valIndWgts.size();
        int count = len;
        int id = 0;
        double[] den = new double[count];
        double accumulateDen = 0;
        StringBuilder sb = new StringBuilder();
//        System.out.print("Subgraph densities:");
        while(count > 0) {
            double density = 0;
            density = computeDensity(valIndWgts,count,sb);
            den[id++] = density;
            accumulateDen += density;
            discardVals.add(sb.toString());
//            System.out.println(sb.toString());
            double min = Double.MAX_VALUE;
            int mid = -1;
            for(int i = 0; i < valIndWgts.size(); i++) {
                double w = valIndWgts.get(i);
                if(w < min) {
                    min = w;
                    mid = i;
                }
            }
            removeOneFeatureValue(mid);
            String str = valIdentifier.remove(mid);
            sb.append(str + ",");
            String[] s = str.split("_");
            int featID = Integer.valueOf(s[0]);
            int valueID = Integer.valueOf(s[1]);
            ValueNode cen = cpList.get(featID).getCenList().get(valueID);
            cen.setWeight(accumulateDen);
            count--;
            
        }
//        System.out.println();
        return den;
    }
    
    
    /**
     * to remove one value from the value candidates
     * @param vid the id of the value to be removed
     */
    public void removeOneFeatureValue(int vid) {
        vMatrix.remove(vid);
        valIndWgts.remove(vid); //virtually remove the value
//            double fWgt = featIndWgts.get(fid);
        for(int k = 0; k < vMatrix.size(); k++) {
            ArrayList<Double> col = vMatrix.get(k);
            double vWgt = valIndWgts.get(k);
            double w = col.remove(vid);
//                double w = col.get(fid);
            valIndWgts.set(k, vWgt-w);
        }
    }
    
    /**
     * to compute the subgraph density using feature level array-list <code>featIndWgts</code>, i.e., average weight per node
     * @param edgeWeights the total incoming edge weights of individual feature values
     * @param valNum the number of values left
     * @param sb to store the non-relevant value ids
     * @return the subgraph density
     */
    public double computeDensity(ArrayList<Double> edgeWeights, int valNum, StringBuilder sb) {
        double density = 0;
        int len = edgeWeights.size();
        for(int i = 0; i < len; i++ ) {
            double w = edgeWeights.get(i);
            density += w;
        }
        density = density / (2*valNum);
//        System.out.print((new DecimalFormat("#0.000000")).format(density)+",");
        return density;
    }
    
    /**
     * to iteratively select top k relevant features
     * @param data a given data set
     * @param featNum the number of feature to be retained
     * @param path the directory for saving the newly generated data sets
     * @param name name of the given data set
     */
    public void featureSelection(Instances data, int featNum, String path, String name){
        Integer[] attrIDArray = new Integer[dim];
        
        for(int i = 0; i < dim; i++) {
            attrIDArray[i] = i;
        }
//        System.out.println();
        sortAndKeepIndex(attrIDArray);
        int len = attrIDArray.length;
        for(int i = 0; i < len; i++) {
            System.out.println(attrIDArray[i]+":"+attrWeight[attrIDArray[i]]+",");
        }
        System.out.println();
        
        int count = len;
        BufferedWriter writer = null;
        StringBuilder sb = new StringBuilder();
        Instances newData;
        int i=0;
        while(count > featNum) { //for a given fixed number
            int id = attrIDArray[i++]+1; //The Remove class assume index starts with 1 rather than 0.
//            System.out.print(id+","+count);
            sb.append(id).append(",");
            count--;
        }
        
        try {
            Remove remove = new Remove();
            System.out.println(sb.toString());
            remove.setAttributeIndices(sb.toString());
            remove.setInvertSelection(false);
            remove.setInputFormat(data);
            newData = Filter.useFilter(data, remove);
            writer = new BufferedWriter(new FileWriter(path+"\\"+name+"_"+String.format("%02d", count)+".arff"));
            writer.write(newData.toString());
            writer.flush();
            writer.close();
            sb.append(",");
        } catch (IOException ex) {
            Logger.getLogger(DSVL.class.getName()).log(Level.SEVERE, null, ex);
        } catch (Exception ex) {
            Logger.getLogger(DSVL.class.getName()).log(Level.SEVERE, null, ex);
        } finally {
            try {
                writer.close();
            } catch (IOException ex) {
                Logger.getLogger(DSVL.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
    }
    
    
    /**
     * to sort the features according to their weights while at the same time keeping track of their orginal indices
     * @param attrIDArray the original feature id array
     */
    public void sortAndKeepIndex(Integer[] attrIDArray) {
        Arrays.sort(attrIDArray, new Comparator<Integer>() {
            @Override
            public int compare(Integer o1, Integer o2) {
                return Double.compare(attrWeight[o1], attrWeight[o2]);
            }
        });
    }
}

File: files/runtime_scalability/algorithm/SDRW_python/sdrw/dsvl.py
"""
DSVL / SDRW core: intra-value outlierness, value graph, dense subgraph discovery.
Python port of edu.uts.aai.utils.DSVL (Java).
"""

from __future__ import annotations

import math
import os
from typing import Dict, List, Optional

from .dataset import Dataset
from .plot import plot_y_points
from .value_centroid import ValueCentroid
from .value_node import ValueNode


class DSVL:
    def __init__(self, cp_list: List[ValueCentroid]) -> None:
        self.cp_list = cp_list
        self.dim = len(cp_list)
        self.v_matrix: List[List[float]] = []
        self.val_feat_ids: List[int] = []
        self.val_identifier: List[str] = []
        self.val_ind_wgts: List[float] = []
        self.attr_weight: Optional[List[float]] = None

    def dense_subgraph_discovery(self, data_size: int) -> str:
        self.calc_intra_feature_value_outlierness(data_size)
        self.value_adjacency_matrix()
        discard_vals: List[str] = []
        den = self.charikar_greedy_search_for_value_graph(discard_vals)
        max_den = float("-inf")
        max_id = -1
        for i, d in enumerate(den):
            if d > max_den:
                max_den = d
                max_id = i
        print(f"MAX:{max_den:.6f} ")
        from . import dsvl4od_utils

        plot_y_points(
            den,
            3,
            dsvl4od_utils.data_set_name,
            dsvl4od_utils.data_set_name,
            "Iteration",
            "Avg. Incoming Edge Weight",
        )
        return discard_vals[max_id]

    def value_outlierness_learning(self, data_size: int) -> None:
        self.calc_intra_feature_value_outlierness(data_size)
        self.value_adjacency_matrix()
        discard_vals: List[str] = []
        self.charikar_greedy_search_for_value_graph(discard_vals)
        self.weighted_degree2_score()

    def calc_intra_feature_value_outlierness(self, data_size: int) -> None:
        m_freq = [0.0] * self.dim
        for i, cp in enumerate(self.cp_list):
            max_freq = 0.0
            for j, cen in enumerate(cp.cen_list):
                global_freq = cen.global_freq(i, j)
                if global_freq > max_freq:
                    max_freq = global_freq
            m_freq[i] = max_freq

        for i, cp in enumerate(self.cp_list):
            for j, cen in enumerate(cp.cen_list):
                global_freq = cen.global_freq(i, j)
                if global_freq == 0:
                    continue
                intra = (
                    abs(global_freq - m_freq[i]) / m_freq[i]
                    + (1 - m_freq[i] / data_size)
                ) / 2.0
                cen.set_intra_od(intra)

    def value_adjacency_matrix(self) -> None:
        print()
        for i in range(self.dim):
            cp = self.cp_list[i]
            for j, cen in enumerate(cp.cen_list):
                if cen.global_freq(i, j) == 0:
                    print("0.0 ", end="")
                    continue

                col: List[float] = []
                tmp = 0.0
                for k in range(self.dim):
                    ai = cen.attr_list[k]
                    gai = cp.get_global_centroid().attr_list[k]
                    length = ai.num_of_value()
                    for l in range(length):
                        if k == cen.org_feat and gai.value(l) != 0:
                            col.append(0.0)
                            print("0.0 ", end="")
                            continue

                        freq = ai.value(l)
                        g_freq = gai.value(l)
                        cen_freq = cen.global_freq(i, j)
                        if cen_freq != 0 and g_freq != 0:
                            w = (
                                cen.get_intra_od()
                                * self.cp_list[k].cen_list[l].get_intra_od()
                                * (freq / (cen_freq * g_freq))
                            )
                            col.append(w)
                            tmp += w
                            print(f"{w} ", end="")
                        else:
                            print("0.0 ", end="")

                self.v_matrix.append(col)
                self.val_feat_ids.append(i)
                self.val_ind_wgts.append(tmp)
                self.val_identifier.append(f"{i}_{j}")
                cen.set_weighted_degree(tmp)
                print()

    def update_weighted_degree(self) -> None:
        for i in range(self.dim):
            cp = self.cp_list[i]
            for j, cen in enumerate(cp.cen_list):
                if cen.global_freq(i, j) == 0:
                    continue
                tmp = 0.0
                for k in range(self.dim):
                    ai = cen.attr_list[k]
                    gai = cp.get_global_centroid().attr_list[k]
                    length = ai.num_of_value()
                    for l in range(length):
                        if k == cen.org_feat and gai.value(l) != 0:
                            continue
                        freq = ai.value(l)
                        g_freq = gai.value(l)
                        cen_freq = cen.global_freq(i, j)
                        if cen_freq != 0 and g_freq != 0:
                            w = (
                                cen.get_weight()
                                * self.cp_list[k].cen_list[l].get_weight()
                                * (freq / (cen_freq * g_freq))
                            )
                            tmp += w
                cen.weighted_degree = tmp

    def weighted_degree2_score(self) -> None:
        self.weight_normalization()
        self.update_weighted_degree()

        self.attr_weight = [1.0] * self.dim
        for i in range(self.dim):
            cp = self.cp_list[i]
            for j, cen in enumerate(cp.cen_list):
                prob = cen.get_weight()
                cen.set_outlier_score(prob)
                print(f"{prob},", end="")
                self.attr_weight[i] *= 1 - prob

        print()
        self.attr_weight = [1 - w for w in self.attr_weight]

    def weight_normalization(self) -> None:
        total_wgt = sum(cen.get_weight() for cp in self.cp_list for cen in cp.cen_list)

        for cp in self.cp_list:
            for cen in cp.cen_list:
                cen.set_weight(cen.get_weight() / total_wgt if total_wgt else 0.0)

    def scoring_test_instances(self, data: Dataset) -> Dict[int, float]:
        scores: Dict[int, float] = {}
        assert self.attr_weight is not None

        for j in range(data.num_instances):
            inst = data.instance(j)
            score = 1.0
            for i in range(self.dim):
                cp = self.cp_list[i]
                index = int(inst.value(i))
                cen = cp.cen_list[index]
                s = cen.get_outlier_score()
                score *= math.pow(1 - s, self.attr_weight[i])
            score = 1 - score
            print(score)
            scores[j] = score

        return scores

    def charikar_greedy_search_for_value_graph(
        self, discard_vals: List[str]
    ) -> List[float]:
        count = len(self.val_ind_wgts)
        edge_weights = list(self.val_ind_wgts)
        v_matrix = [list(row) for row in self.v_matrix]
        identifiers = list(self.val_identifier)

        den: List[float] = []
        id_ = 0
        accumulate_den = 0.0
        sb: List[str] = []

        while count > 0:
            density = self._compute_density(edge_weights, count)
            den.append(density)
            id_ += 1
            accumulate_den += density
            discard_vals.append(",".join(sb))

            min_w = float("inf")
            mid = -1
            for i, w in enumerate(edge_weights):
                if w < min_w:
                    min_w = w
                    mid = i

            self._remove_one_feature_value(mid, edge_weights, v_matrix)
            str_id = identifiers.pop(mid)
            sb.append(f"{str_id},")
            feat_id, value_id = (int(x) for x in str_id.split("_"))
            cen = self.cp_list[feat_id].cen_list[value_id]
            cen.set_weight(accumulate_den)
            count -= 1

        return den

    def _remove_one_feature_value(
        self,
        vid: int,
        edge_weights: List[float],
        v_matrix: List[List[float]],
    ) -> None:
        v_matrix.pop(vid)
        edge_weights.pop(vid)
        for k in range(len(v_matrix)):
            col = v_matrix[k]
            w = col.pop(vid)
            edge_weights[k] -= w

    @staticmethod
    def _compute_density(edge_weights: List[float], val_num: int) -> float:
        density = sum(edge_weights)
        return density / (2 * val_num) if val_num else 0.0

    def feature_selection(
        self, data: Dataset, feat_num: int, path: str, name: str
    ) -> None:
        attr_id_array = list(range(self.dim))
        assert self.attr_weight is not None
        attr_id_array.sort(key=lambda idx: self.attr_weight[idx])

        for i, attr_id in enumerate(attr_id_array):
            print(f"{attr_id}:{self.attr_weight[attr_id]},")
        print()

        count = len(attr_id_array)
        remove_ids: List[int] = []
        i = 0
        while count > feat_num:
            remove_ids.append(attr_id_array[i] + 1)
            i += 1
            count -= 1

        new_data = _remove_attributes(data, remove_ids)
        out_path = os.path.join(path, f"{name}_{count:02d}.arff")
        _write_arff(new_data, out_path)
        print(",".join(str(x) for x in remove_ids))


def _remove_attributes(data: Dataset, one_based_indices: List[int]) -> Dataset:
    """Remove attributes by 1-based indices (Weka Remove convention)."""
    remove_set = {i - 1 for i in one_based_indices}
    keep = [i for i in range(data.num_attributes()) if i not in remove_set]
    attrs = [data.attributes[i] for i in keep]
    matrix = data.data[:, keep].copy()
    result = Dataset(attrs, matrix, class_index=-1)
    if data.class_index in keep:
        result.set_class_index(keep.index(data.class_index))
    elif len(keep) > 0:
        result.set_class_index(len(keep) - 1)
    return result


def _write_arff(data: Dataset, path: str) -> None:
    import os

    lines = ["@relation SDRW_subset", ""]
    for attr in data.attributes:
        if attr.is_nominal():
            values = ",".join(f'"{v}"' for v in attr.values)
            lines.append(f"@attribute {attr.name} {{{values}}}")
        else:
            lines.append(f"@attribute {attr.name} numeric")
    lines.append("")
    lines.append("@data")
    for row in data.data:
        parts = []
        for i, val in enumerate(row):
            if data.attributes[i].is_nominal():
                parts.append(data.attributes[i].values[int(val)])
            else:
                parts.append(str(val))
        lines.append(",".join(parts))

    os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
    with open(path, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))
